BUG: need better inference for path in Series construction (GH9456)

patrickfournier opened the pull request:
closes  #9456.

Commit:
TST: Adding test for bug GH 9456

Commit:
BUG: GH 9456 Fixed Series.__init__ to better handle dict data

- dict with datetime64 keys now working
- when isinstance(index, DatetimeIndex), use lib.fast_multiget correctly
  to avoid raising a TypeError exception

jreback commented on pandas/core/series.py:
```diff
@@ -166,9 +166,9 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
                         index = Index(_try_sort(data))
                 try:
-                    if isinstance(index, DatetimeIndex):
+                    if isinstance(index, DatetimeIndex) and lib.infer_dtype(data) != 'datetime64':
```
see why this path is hit in the first place and can go from there.

patrickfournier replied:
Before I added `.value` at line 171, `lib.fast_multiget` was throwing a `TypeError`. Now, the `else` clause must handles the conversion of datetime64 dicts because we cannot compare a `Timestamp` with a `datetime64`:

```python
>>> d = {datetime.datetime(2015, 1, 7, 2, 0): 42544017.198965244}
>>> i = pandas.core.index.Index(d)
>>> i.astype('O').values
array([Timestamp('2015-01-07 02:00:00')], dtype=object)
>>> datetime.datetime(2015, 1, 7, 2, 0) == pandas.lib.Timestamp('2015-01-07 02:00:00')
True

>>> d = {numpy.datetime64('2015-01-06T19:00:00.000000000-0500'): 42544017.198965244}
>>> i = pandas.core.index.Index(d)
>>> i.astype('O').values
array([Timestamp('2015-01-07 00:00:00')], dtype=object)
>>> numpy.datetime64('2015-01-06T19:00:00.000000000-0500') == pandas.lib.Timestamp('2015-01-07 00:00:00')
False
```

jreback replied:
that's not what I mean. put halt in there and see what test actually hits this. This is handling an edge case. Your fix, just confuses things. I think there is a more general soln.

patrickfournier replied:
I ran the tests and reported the results in the issue: https://github.com/pydata/pandas/issues/9456.

Maybe a better fix would be to correctly handle the comparison between Timestamp and datetime64?

jreback replied:
The idea is to have the least special cases as possible. You rarely actually want to convert a ``DatetimeIndex`` using ``.astype('O')`` this is the least desirable result (though it may have been done for some reason). But it is tricky because a user can pass ``Timestamp/datetime.datetime/datetime.date/np.datetime64`` as elements and a non-index like (but that is index-like). Best best is to do ``_ensure_index`` which will convert list-like things to an ``Index`` (and possibly to a ``DatetimeIndex`` if its compat; if its not then all bets are off).

Commit:
Merge branch 'master' into issue-9456

Commit:
BUG: GH 9456 Fixed Series.__init__ to better handle dict data

- dict with datetime64 keys now working
- use lib.fast_multiget correctly to avoid unnecessary exceptions

patrickfournier commented:
I rewrote the fix to be more generic. If the type of Index elements is the same as the type of data elements, we use the Index to access and copy the dict value. If not, we use the original index. See comments in diff.

patrickfournier commented on pandas/core/series.py:
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
-                        data = [data.get(i, nan) for i in index]
+                    # lib.fast_multiget raises TypeError if type(data) != dict
+
+                    if lib.infer_dtype(data) == lib.infer_dtype(index.values):
```
Here we try to avoid a call to `np.array()` by checking if we can access `data` values using `index.values`. However I checked the code for `infer_dtype` and it may be quite complex. Maybe we should just go straight to the `else`?

patrickfournier commented on pandas/core/series.py:
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
+                    index = Index(original_index)
```
I wanted to use `_ensure_index`, but it does not create a `MultiIndex` when the dict keys are tuples.

patrickfournier commented on pandas/core/series.py:
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
+                        else:
+                            # np.array(['z', ('a', 'b')]) raises ValueError;
+                            # this may happens with MultiIndex.
+                            data = lib.fast_multiget(data, np.array(original_index),
```
If we `lib.fast_multiget` supported `list` as its second parameter, we could save the call to `np.array`.

jreback commented on pandas/core/series.py:
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
                     else:
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
```
why is it necessary to use a new variable here. This original code looks fine.

patrickfournier replied:
Since we cannot always reorder the `dict` using the `Index` (because of incompatible data types), I keep the original index to use it if I cannot use the `Index`.

jreback commented:
this overlaps with #10269

jreback commented:
superseded by #10269 

thanks!
